### PR TITLE
Supersed pylint_allowlist

### PR DESCRIPTION
**Pull request by bhack**
This a first PR to try to supersed `pylint_allowlist` and custom pattern matching scripts for a less critical integration of our `pylint` linting with third party tools (e.g. IDE) and enable a potential pre-commit hook that we want to officially distribute for linting.

I've upgraded  `.pylintrc` with the info that we currently have in the pattern matching:
https://github.com/tensorflow/tensorflow/blob/38f4134e6884fb1d766999486a72810ebae6acf5/tensorflow/tools/ci_build/ci_sanity.sh#L181-L198

Then I used [pylint-silent](https://pypi.org/project/pylint-silent/) to patch files.

### CODE REVIEW

**Comment by bhack** [comment]
We have now 14 residual `pylint` linting errors that I cannot reproduce in our official Docker devel image (for the command see https://github.com/tensorflow/tensorflow/pull/48291).
What is the Ubuntu Sanity image?

Do we want to use the official devel image in the CI to be sure that the user/contributors dev environment is aligned with the CI Sanity pass in github?

File: tensorflow/compiler/tests/image_ops_test.py [hunk 1]
```diff
@@ -91,13 +91,13 @@ def testRGBToHSVNumpy(self):
...
       rgb_flat = _generate_numpy_random_rgb((64, 3)).astype(nptype)
-      rgb_np = rgb_flat.reshape(4, 4, 4, 3)
+      rgb_np = rgb_flat.reshape(4, 4, 4, 3)  # pylint: disable=too-many-function-args
       hsv_np = np.array([
           colorsys.rgb_to_hsv(
               r.astype(np.float64), g.astype(np.float64), b.astype(np.float64))
           for r, g, b in rgb_flat
       ])
-      hsv_np = hsv_np.reshape(4, 4, 4, 3)
+      hsv_np = hsv_np.reshape(4, 4, 4, 3)  # pylint: disable=too-many-function-args
```
Comment (bhack) [inline tensorflow/compiler/tests/image_ops_test.py:100]: E.g. We are tracking this type of notifications with https://github.com/tensorflow/tensorflow/issues/41873

**Review by mdanatg — APPROVED** [review]
This is good, it gives us a better indication of where the flagged things are. Cleaning them up one by one (were possible), would make useful future contributions.

**Comment by gbaned** [comment]
@bhack  Can you please address Ubuntu Sanity errors? Thanks!

**Comment by bhack** [comment]
I need to understand what Is the referenc Sanity Docker Image. 
I prefer to have in CI our official Docker devel image so that we are sure to have the same CI and contributor/developer experience

**Review by mihaimaruseac — APPROVED** [review]

**Comment by bhack** [comment]
@mihaimaruseac What is the Docker image used by CI Ubuntu Sanity?

**Comment by bhack** [comment]
In the mean time I've also upgraded the Devel images to run the python linting step at https://github.com/tensorflow/tensorflow/pull/48371

**Comment by bhack** [comment]
I cannot reproduce the errors we have in the Sanity CI logs  locally also forcing the same Python version. E.g.
E.g. in the CI log we have:
`tensorflow/python/training/tracking/python_state_test.py:131: [E1123(unexpected-keyword-arg), _NumpyWrapper.deserialize] Unexpected keyword argument 'allow_pickle' in function call`

But

```
docker run --rm -it -v $PWD:/tensorflow -w /tensorflow python:3.8.9-buster bash -c  "pip install pylint && pylint tensorflow/python/training/tracking/python_state_test.py"
```
```
------------------------------------
Your code has been rated at 10.00/10
```

**Comment by mihaimaruseac** [comment]
I don't think we build using Docker at the moment.

**Comment by bhack** [comment]
> I don't think we build using Docker at the moment.

Ok but I need to align the env locallly to have the same error as in CI

**Comment by bhack** [comment]
I've changed almost every `install numpy==1.14.5` string but the CI is still on `numpy 1.14.5`

File: tensorflow/tools/ci_build/ci_sanity.sh [hunk 2]
```diff
@@ -129,6 +129,12 @@ do_pylint() {
   echo ""
   echo "check whether pylint is available or not."
   echo ""
+
+  # !!!!!!This is only an hack to workaround CI VM with numpy 1.14.5!!!!!!!!
+  python3.8 -m pip install --upgrade numpy~=1.19.2 lazy-object-proxy
+  python3.8 -m pip show numpy lazy-object-proxy
+
```
Comment (bhack) [inline tensorflow/tools/ci_build/ci_sanity.sh:136]: I've explicitly temp forced numpy for python3.8 here but if you see the log we have still some residual misalignment with the CI VM

```
FAIL: Found 4 errors
tensorflow/python/data/experimental/kernel_tests/data_service_ops_ft_test.py:109: [W0622(redefined-builtin), DataServiceOpsTest.testDispatcherRestartDuringDistributedEpochRepeat] Redefining built-in 'breakpoint'
tensorflow/python/keras/preprocessing/image.py:305: [E0240(inconsistent-mro), DirectoryIterator] Inconsistent method resolution order for class 'DirectoryIterator'
tensorflow/python/ops/control_flow_grad.py:31: [W0622(redefined-builtin), ] Redefining built-in 'exit'
tensorflow/python/ops/control_flow_grad.py:31: [W0622(redefined-builtin), ] Redefining built-in 'tuple'
```

**Comment by gbaned** [comment]
@mihaimaruseac Can you please take a look on the above comment from @bhack. Thanks!

**Comment by bhack** [comment]
I've opened https://github.com/tensorflow/build/pull/27 in the SIG-build repository so that they can compare `ci_sanity.sh` VS `pylint` results on this branch with Github Action.

**Comment by gbaned** [comment]
@mihaimaruseac Can you please take a look on the above comments from @bhack. Thanks!

**Comment by bhack** [comment]
@gbaned I've already talked with @mihaimaruseac just yesterday and he his still busy.

**Comment by bhack** [comment]
Also we are waiting for a comment at https://github.com/tensorflow/build/pull/27#issuecomment-827819407 for the Ubuntu Sanity VM.

**Comment by angerson** [comment]
I checked the environment for the machine that's running ci_sanity (you can actually see it in ci_sanity results for new PRs):

```
absl-py==0.8.1
appdirs==1.4.3
astroid==2.5.6
attrs==19.3.0
cachetools==3.1.1
certifi==2020.12.5
chardet==4.0.0
distlib==0.3.0
filelock==3.0.12
future==0.18.2
gast==0.2.2
google-auth==1.9.0
google-auth-oauthlib==0.4.1
grpcio==1.25.0
h5py==2.10.0
idna==2.10
isort==5.8.0
joblib==0.14.1
Keras-Preprocessing==1.1.0
lazy-object-proxy==1.6.0
Markdown==3.1.1
mccabe==0.6.1
mock==3.0.5
numpy==1.14.5
oauthlib==3.1.0
portpicker==1.3.1
protobuf==3.11.1
pyasn1==0.4.8
pyasn1-modules==0.2.7
pycurl==7.43.0
pygobject==3.20.0
pylint==2.7.2
python-apt==1.1.0b1+ubuntu0.16.4.11
PyYAML==3.11
requests==2.25.1
requests-oauthlib==1.3.0
rsa==4.0
scikit-learn==0.22
scipy==1.4.0
screen-resolution-extra==0.0.0
six==1.12.0
tb-nightly==2.1.0a20191106
tf-estimator-nightly==2.0.0.dev2019121609
toml==0.10.2
typing==3.7.4.1
unattended-upgrades==0.1
urllib3==1.26.4
virtualenv==20.0.5
Werkzeug==0.16.0
wrapt==1.12.1
xkit==0.0.0

check whether pylint is available or not.

pylint 2.7.2
astroid 2.5.6
Python 3.8.9 (default, Apr  3 2021, 01:02:10)
[GCC 5.4.0 20160609]
```

**Comment by bhack** [comment]
The last remaining issue for this PR is how we could align a Little bit that numpy version with the version currently in our official Docker devel  https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/dockerfiles/dockerfiles/devel-cpu.Dockerfile#L87

Cause I don't know what reference env to suggest to contributors for local linting on par with CI Sanity.

**Comment by gbaned** [comment]
Seems auto-merge is not happening but the changes are merged into master now, so we can close this. Thank you for the PR.

**Comment by angerson** [comment]
Thank you @gbaned. I had to make a number of extra changes internally to land this, but it's in now.

**Comment by bhack** [comment]
@angerson Thanks. Let see if we could review https://github.com/tensorflow/tensorflow/pull/48371 so that we have a reference local env for developers to run reproducible liniting.

**Comment by bhack** [comment]
It was merged but the Ubuntu sanity VM is still on `numpy==1.14.5`. As I told in https://github.com/tensorflow/tensorflow/pull/48294#issuecomment-843620546 this will fail the sanity check in the CI.

**Pull request closed without merging**